File: validate.py
"""
Validate module - Validate elevation data
"""
import json


class ElevationValidator:
    """Validate elevation data"""
    
    def __init__(self, min_elevation=0, max_elevation=2600):
        """
        Initialize validator
        
        Args:
            min_elevation: Minimum valid elevation in meters (Romania: 0 = Black Sea)
            max_elevation: Maximum valid elevation in meters (Romania: ~2544m Moldoveanu)
        """
        self.min_elevation = min_elevation
        self.max_elevation = max_elevation
    
    def validate_element(self, element):
        """
        Validate elevation for a single element
        
        Returns:
            dict with validation results
        """
        result = {
            "valid": False,
            "element_id": element.get("id"),
            "element_type": element.get("type"),
            "elevation": None,
            "errors": []
        }
        
        # Check if elevation exists
        if "elevation_fetched" not in element:
            result["errors"].append("No elevation data")
            return result
        
        elevation = element["elevation_fetched"]
        result["elevation"] = elevation
        
        # Validate range
        if elevation < self.min_elevation:
            result["errors"].append(f"Elevation {elevation}m below minimum {self.min_elevation}m")
        elif elevation > self.max_elevation:
            result["errors"].append(f"Elevation {elevation}m above maximum {self.max_elevation}m")
        else:
            result["valid"] = True
        
        return result
    
    def validate_elements(self, elements):
        """
        Validate multiple elements
        
        Returns:
            dict with valid and invalid elements
        """
        valid = []
        invalid = []
        
        for element in elements:
            validation = self.validate_element(element)
            
            if validation["valid"]:
                valid.append(element)
            else:
                invalid.append({
                    "element": element,
                    "validation": validation
                })
        
        return {
            "valid": valid,
            "invalid": invalid
        }
    
    def validate_all(self, data):
        """Validate all categories of data"""
        results = {}
        
        for category, elements in data.items():
            if elements:
                print(f"\nValidating {category}...")
                validation = self.validate_elements(elements)
                results[category] = validation
                
                print(f"  Valid: {len(validation['valid'])}")
                print(f"  Invalid: {len(validation['invalid'])}")
                
                # Show invalid examples
                if validation['invalid']:
                    print(f"  Invalid examples:")
                    for item in validation['invalid'][:3]:
                        val = item['validation']
                        print(f"    - ID {val['element_id']}: {', '.join(val['errors'])}")
        
        return results


if __name__ == "__main__":
    # Load enriched data
    with open("output/osm_data_enriched.json", "r", encoding="utf-8") as f:
        data = json.load(f)
    
    # Validate
    validator = ElevationValidator(min_elevation=0, max_elevation=2600)
    results = validator.validate_all(data)
    
    # Save validation results
    output = {}
    for category, validation in results.items():
        output[category] = {
            "valid_count": len(validation["valid"]),
            "invalid_count": len(validation["invalid"]),
            "valid_elements": validation["valid"]
        }
    
    with open("output/osm_data_validated.json", "w", encoding="utf-8") as f:
        json.dump(output, f, indent=2, ensure_ascii=False)
    
    print("\nValidation complete! Results saved to output/osm_data_validated.json")

File: src/validate.go
package main

import (
	"fmt"
)

type ElevationValidator struct {
	MinElevation float64
	MaxElevation float64
}

type ValidationResult struct {
	Valid       bool     `json:"valid"`
	ElementID   int64    `json:"element_id"`
	ElementType string   `json:"element_type"`
	Elevation   *float64 `json:"elevation"`
	Errors      []string `json:"errors"`
}

type ValidationResults struct {
	Valid   []OSMElement     `json:"valid"`
	Invalid []InvalidElement `json:"invalid"`
}

type InvalidElement struct {
	Element    OSMElement       `json:"element"`
	Validation ValidationResult `json:"validation"`
}

type ValidatedCategory struct {
	ValidCount    int          `json:"valid_count"`
	InvalidCount  int          `json:"invalid_count"`
	ValidElements []OSMElement `json:"valid_elements"`
}

type ValidatedData struct {
	TrainStations       ValidatedCategory `json:"train_stations"`
	AlpineHuts          ValidatedCategory `json:"alpine_huts"`
	OtherAccommodations ValidatedCategory `json:"other_accommodations"`
}

func NewElevationValidator(minElevation, maxElevation float64) *ElevationValidator {
	return &ElevationValidator{
		MinElevation: minElevation,
		MaxElevation: maxElevation,
	}
}

func (v *ElevationValidator) ValidateElement(element OSMElement) ValidationResult {
	result := ValidationResult{
		Valid:       false,
		ElementID:   element.ID,
		ElementType: element.Type,
		Elevation:   nil,
		Errors:      []string{},
	}

	// Check if elevation exists
	if element.ElevationFetched == nil {
		result.Errors = append(result.Errors, "No elevation data")
		return result
	}

	elevation := *element.ElevationFetched
	result.Elevation = &elevation

	// Validate range
	if elevation < v.MinElevation {
		result.Errors = append(result.Errors,
			fmt.Sprintf("Elevation %.1fm below minimum %.1fm", elevation, v.MinElevation))
	} else if elevation > v.MaxElevation {
		result.Errors = append(result.Errors,
			fmt.Sprintf("Elevation %.1fm above maximum %.1fm", elevation, v.MaxElevation))
	} else {
		result.Valid = true
	}

	return result
}

func (v *ElevationValidator) ValidateElements(elements []OSMElement) ValidationResults {
	results := ValidationResults{
		Valid:   []OSMElement{},
		Invalid: []InvalidElement{},
	}

	for _, element := range elements {
		validation := v.ValidateElement(element)

		if validation.Valid {
			results.Valid = append(results.Valid, element)
		} else {
			results.Invalid = append(results.Invalid, InvalidElement{
				Element:    element,
				Validation: validation,
			})
		}
	}

	return results
}

func (v *ElevationValidator) ValidateAll(data *EnrichedData) map[string]ValidationResults {
	results := make(map[string]ValidationResults)

	categories := map[string][]OSMElement{
		"train_stations":       data.TrainStations,
		"alpine_huts":          data.AlpineHuts,
		"other_accommodations": data.OtherAccommodations,
	}

	for category, elements := range categories {
		if len(elements) > 0 {
			fmt.Printf("\nValidating %s...\n", category)
			validation := v.ValidateElements(elements)
			results[category] = validation

			fmt.Printf("  Valid: %d\n", len(validation.Valid))
			fmt.Printf("  Invalid: %d\n", len(validation.Invalid))

			// Show invalid examples
			if len(validation.Invalid) > 0 {
				fmt.Println("  Invalid examples:")
				for i, item := range validation.Invalid {
					if i >= 3 {
						break
					}
					val := item.Validation
					fmt.Printf("    - ID %d: %v\n", val.ElementID, val.Errors)
				}
			}
		}
	}

	return results
}

func runValidate() error {
	fmt.Println("\n" + string(repeat('=', 60)))
	fmt.Println("STEP 4: VALIDATE - Checking elevation ranges (0-2600m)")
	fmt.Println(string(repeat('=', 60)))

	// Load enriched data
	var data EnrichedData
	if err := loadJSON("output/osm_data_enriched.json", &data); err != nil {
		return fmt.Errorf("output/osm_data_enriched.json not found. Run --enrich first: %v", err)
	}

	// Validate
	validator := NewElevationValidator(0, 2600)
	results := validator.ValidateAll(&data)

	// Save validation results
	output := ValidatedData{
		TrainStations: ValidatedCategory{
			ValidCount:    len(results["train_stations"].Valid),
			InvalidCount:  len(results["train_stations"].Invalid),
			ValidElements: results["train_stations"].Valid,
		},
		AlpineHuts: ValidatedCategory{
			ValidCount:    len(results["alpine_huts"].Valid),
			InvalidCount:  len(results["alpine_huts"].Invalid),
			ValidElements: results["alpine_huts"].Valid,
		},
		OtherAccommodations: ValidatedCategory{
			ValidCount:    len(results["other_accommodations"].Valid),
			InvalidCount:  len(results["other_accommodations"].Invalid),
			ValidElements: results["other_accommodations"].Valid,
		},
	}

	if err := saveJSON("output/osm_data_validated.json", output); err != nil {
		return err
	}

	fmt.Println("\n✓ Validation complete! Results saved to output/osm_data_validated.json")

	return nil
}
